カレンダー表示プログラムを作成

triton27 opened the pull request.

triton27 committed:
first commit

triton27 committed:
add: カレンダー出力プログラムを追加

triton27 committed:
refactor: 変数名の変更、メソッドの切り分けなど

triton27 committed:
add: minimistライブラリの追加

triton27 committed:
fix: package.json に type: module を追加して ES モジュールをサポート

triton27 committed:
fix: ESLint対応

エラーメッセージ↓
  1:12  error  'require' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)

triton27 committed:
fix: Prettierによるフォーマット修正

triton27 committed:
fix: Prettierによるフォーマット修正

triton27 committed:
delete: 廃止されたフォーマットルールを削除

cafedomancer commented on 02.calendar/prettier.config.js:
Prettier の設定は変更しないてください 🙏

triton27 replied:
3d920ca で戻しました。

cafedomancer commented on 02.calendar/package.json:
プラクティスの要件を満たすにあたって必要のない変更が多く含まれています。

triton27 replied:
3d920ca で戻しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+import minimist from 'minimist';
+
+var argv = minimist(process.argv.slice(2));
```
`var` は古い変数宣言の仕組みなので使ってはいけません。`let` や `const` を使いましょう。

triton27 replied:
9ea211c で修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+
+var argv = minimist(process.argv.slice(2));
+const now = new Date();
+const year = argv['y'] || now.getFullYear();
+const month = argv['m'] || now.getMonth() + 1;
```
オブジェクトのプロパティにアクセスするときには、ブラケット記法でないとアクセスできないようなプロパティ名である場合を除いて、ドット記法を使ってください。また、`||` 演算子だと左辺が `0` の場合にも右辺が評価されてしまいます。別の演算子を使ってください。

triton27 replied:
fbed5e6 で修正しました。

ご指摘ありがとうございます。
プロパティ名が数値や記号を含むなど特殊な条件でない限り、ドット記法を使用するべきなんですね。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+const year = argv['y'] || now.getFullYear();
+const month = argv['m'] || now.getMonth() + 1;
+
+const SPACE = ' ';
```
これはリテラルを見れば分かるので定数に入れる必要はないと思います。

triton27 replied:
b521a9f で修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+import minimist from 'minimist';
+
+var argv = minimist(process.argv.slice(2));
+const now = new Date();
+const year = argv['y'] || now.getFullYear();
+const month = argv['m'] || now.getMonth() + 1;
```
main 相当の処理と関数定義は混在させて書かないでください。関数定義、main 相当の処理、の順に記述してください。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+const SPACE = ' ';
+
+function printCalendarHeader(year, month) {
+  const weekDays = ['日', '月', '火', '水', '木', '金', '土'];
```
この内容を配列として表現する必要はないと思います。出力したい文字列をそのまま出力すれば良いだけだからです。

triton27 replied:
e170f20 で修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+  }
+}
+
+function getDayOfWeek(year, month) {
+  return new Date(year, month - 1, 1).getDay();
+}
+
+function getTotalDays(year, month) {
+  return new Date(year, month, 0).getDate();
+}
```
関数にする必要のない処理だと思います。オブジェクトからそのままデータを取得すれば良いからです。https://bootcamp.fjord.jp/pages/utilize-objects

triton27 replied:
1380ab6 で修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+  process.stdout.write(`${SPACE.repeat(3).repeat(dayOfWeek)}`);
+
+  for (let day = 1; day <= totalDays; day++) {
+    let dayString = day.toString().padStart(2, SPACE) + SPACE;
```
文字列結合ではなくテンプレートを文字列を使ってください。また、この実装だと土曜日の日付の後ろや月末の日付の後ろにも余分な空白が追加されてしまうと思います。

triton27 replied:
1aedf85 で修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+
+  for (let day = 1; day <= totalDays; day++) {
+    let dayString = day.toString().padStart(2, SPACE) + SPACE;
+    if (dayOfWeek == 6) {
```
等価演算子ではなく厳密等価演算子を使ってください。

triton27 replied:
675f46d で修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,46 @@
+      dayOfWeek = 0;
+    } else {
+      process.stdout.write(dayString);
+      dayOfWeek++;
```
`day` と `dayOfWeek` という 2 つのループカウンタがあるのは分かりづらいと思います。Date オブジェクトをうまく活用すればこのループカウンタは不要になるはずです。

triton27 replied:
1aedf85 で修正しました。

triton27 committed:
fix: 変数宣言キーワードを修正

triton27 committed:
fix: 不要な定数を削除し、空白文字を直接記述

triton27 committed:
fix: 不要な変数を削除し、曜日を直接記述

triton27 committed:
fix: 厳密な比較に変更

triton27 committed:
fix: 不要な修正を戻す

triton27 committed:
fix: 演算子に変更して、undefinedやnullを適切に扱うように修正。argvプロパティへのアクセスをドット記法に統一

triton27 committed:
fix: 不要な関数を削除し、日付計算結果を直接渡す

triton27 committed:
fix: 日付の出力調整
　- 初日の出力位置を調整
　- 土曜日や月末の日付の後に余分な空白が追加されないよう修正
　- 文字列結合をテンプレートリテラルに変更し、可読性を向上
　- day, dayOfWeek の2つのループカウンタを削除し、Dateオブジェクトを使用してロジックを簡素化

triton27 committed:
refactor: ファイル名だけで実行できるようにシバンを追加

triton27 committed:
fix: カレンダー出力後に１行空行を出力

triton27 committed:
fix: 初日が土曜日だった場合、負の値になることを避ける

triton27 committed:
fix: ESLintとPrettierによるフォーマット修正

cafedomancer commented on 02.calendar/prettier.config.js:
このファイルは消さないでください！

triton27 replied:
修正を戻しました、ご確認お願いします。🙇‍♂️

cafedomancer commented on 02.calendar/package.json:
インデントを変更しないでください！

triton27 replied:
修正を戻しました、ご確認お願いします。🙇‍♂️

triton27 committed:
fix: 不要な修正を戻す

triton27 committed:
fix: 不要な修正を戻す

triton27 committed:
fix: インデントを修正

triton27 committed:
add: minimist

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,38 @@
+
+import minimist from "minimist";
+
+const argv = minimist(process.argv.slice(2));
+const now = new Date();
+const year = argv.y ?? now.getFullYear();
+const month = argv.m ?? now.getMonth() + 1;
+
+function printCalendarHeader(year, month) {
+  console.log(`${month}月 ${year}`.padStart(13, " "));
+  console.log(`日 月 火 水 木 金 土`);
+}
+
+function printCalendarDays(year, month) {
+  // 対象年月の1日目の曜日を初期値として取得
+  let firstDayOfWeek = new Date(year, month - 1, 1).getDay();
+  // 対象年月の最終日を取得
+  const totalDays = new Date(year, month, 0).getDate();
+
+  // １日目の出力場所を調整
+  process.stdout.write(`${" ".repeat(Math.max(0, 3 * firstDayOfWeek - 1))}`);
+
+  for (let day = 1; day <= totalDays; day++) {
+    let dayOfWeek = new Date(year, month - 1, day).getDay();
+    if (dayOfWeek === 6) {
+      console.log(day.toString().padStart(3, " "));
+    } else if (dayOfWeek === 0) {
+      process.stdout.write(day.toString().padStart(2, " "));
+    } else {
+      process.stdout.write(day.toString().padStart(3, " "));
+    }
+  }
+  console.log();
+}
+
+printCalendarHeader(year, month);
+printCalendarDays(year, month);
```
https://github.com/triton27/js-practices/pull/2#discussion_r1713108454 の対応が漏れています。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,38 @@
+
+function printCalendarDays(year, month) {
+  // 対象年月の1日目の曜日を初期値として取得
+  let firstDayOfWeek = new Date(year, month - 1, 1).getDay();
```
オブジェクトはオブジェクトのまま扱いましょう。https://bootcamp.fjord.jp/pages/utilize-objects

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+function printCalendarDays(year, month) {
+  // 対象年月の1日目の曜日を初期値として取得
```
コードから分かることをコメントとして残すのはやめましょう。コードとコメントの間に乖離が発生するからです (現に、初期値ではないという点で乖離が発生している) 。コードからは分からないことだけコメントとして残してください。

triton27 replied:
5966bef で修正しました。

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,38 @@
+  const totalDays = new Date(year, month, 0).getDate();
+
+  // １日目の出力場所を調整
+  process.stdout.write(`${" ".repeat(Math.max(0, 3 * firstDayOfWeek - 1))}`);
```
`Math.max(0, 3 * firstDayOfWeek - 1)` の部分が分かりづらい気がするのですが、これはどんな動きをするんでしょうか？

cafedomancer commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,38 @@
+
+  for (let day = 1; day <= totalDays; day++) {
+    let dayOfWeek = new Date(year, month - 1, day).getDay();
+    if (dayOfWeek === 6) {
+      console.log(day.toString().padStart(3, " "));
+    } else if (dayOfWeek === 0) {
+      process.stdout.write(day.toString().padStart(2, " "));
+    } else {
+      process.stdout.write(day.toString().padStart(3, " "));
+    }
```
分岐間で異なる部分のみを分岐として抽出したほうがいいと思います。標準出力への出力や padding の処理が分岐間で重複していて、どんな処理を分岐させたいのかが分かりづらいです。

triton27 replied:
fb6db64(順番整理), 8e034f0(インデント修正) で修正しました。

triton27 committed:
refactor: 関数定義とメイン処理の順番を整理

triton27 committed:
chore: 不要なコメントを削除

triton27 committed:
fix: インデントを修正
